crates/core/src/marzano_context.rs

```diff
@@ -0,0 +1,273 @@
+    pub predicate_definitions: &'a Vec<PredicateDefinition<MarzanoQueryContext>>,
+    pub function_definitions: &'a Vec<GritFunctionDefinition<MarzanoQueryContext>>,
+    pub foreign_function_definitions: &'a Vec<ForeignFunctionDefinition>,
+    pub files: &'a FileOwners,
+    pub built_ins: &'a BuiltIns,
+    pub language: &'a TargetLanguage,
+    pub runtime: &'a ExecutionContext,
+    pub name: Option<String>,
+}
+
+impl<'a> MarzanoContext<'a> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        pattern_definitions: &'a Vec<PatternDefinition<MarzanoQueryContext>>,
+        predicate_definitions: &'a Vec<PredicateDefinition<MarzanoQueryContext>>,
+        function_definitions: &'a Vec<GritFunctionDefinition<MarzanoQueryContext>>,
+        foreign_function_definitions: &'a Vec<ForeignFunctionDefinition>,
+        files: &'a FileOwners,
+        built_ins: &'a BuiltIns,
+        language: &'a TargetLanguage,
+        runtime: &'a ExecutionContext,
+        name: Option<String>,
+    ) -> Self {
+        Self {
+            pattern_definitions,
+            predicate_definitions,
+            function_definitions,
+            foreign_function_definitions,
+            files,
+            built_ins,
+            language,
+            runtime,
+            name,
+        }
+    }
+}
+
+impl<'a> ExecContext<'a, MarzanoQueryContext> for MarzanoContext<'a> {
+    fn pattern_definitions(&self) -> &[PatternDefinition<MarzanoQueryContext>] {
+        self.pattern_definitions
+    }
+
+    fn predicate_definitions(&self) -> &[PredicateDefinition<MarzanoQueryContext>] {
+        self.predicate_definitions
+    }
+
+    fn function_definitions(&self) -> &[GritFunctionDefinition<MarzanoQueryContext>] {
+        self.function_definitions
+    }
+
+    fn foreign_function_definitions(&self) -> &[ForeignFunctionDefinition] {
+        self.foreign_function_definitions
+    }
+
+    fn ignore_limit_pattern(&self) -> bool {
+        self.runtime.ignore_limit_pattern
+    }
+
+    fn call_built_in(
+        &self,
+        call: &'a CallBuiltIn<MarzanoQueryContext>,
+        context: &'a Self,
+        state: &mut State<'a, MarzanoQueryContext>,
+        logs: &mut AnalysisLogs,
+    ) -> Result<MarzanoResolvedPattern<'a>> {
+        self.built_ins.call(call, context, state, logs)
+    }
+
+    #[cfg(all(
+        feature = "network_requests_external",
+        feature = "external_functions_ffi",
+        not(feature = "network_requests"),
+        target_arch = "wasm32"
+    ))]
+    fn exec_external(
+        &self,
+        code: &[u8],
+        param_names: Vec<String>,
+        input_bindings: &[&str],
+    ) -> Result<Vec<u8>> {
+        (self.runtime.exec_external)(code, param_names, input_bindings)
+    }
+
+    // FIXME: Don't depend on Grit's file handling in context.
+    fn files(&self) -> &FileOwners {
+        self.files
+    }
+
+    fn language(&self) -> &TargetLanguage {
+        self.language
+    }
+
+    fn exec_step(
```

Comment:
This trait implementation grew significantly due to the implementatoin of this `exec_step()` method, which was moved from the `Step` pattern. Since I was moving anyway, I put the entire `MarzanoContext` into its own file.

Note how the method above used to have a FIXME, which is now resolved :)